Refactor tests using Theory

Philipp-Binder opened the pull request:
While doing major changes (eg changing from Sprache to Superpower) I had some trouble with testing, because the current Fact-Methods do fail at the first failing case. That way it is hard to check whether you fixed something and just broke a different case, or if it does not fix the intended case at all.
Additionally you always have to care about whether the testcases within a method are really "the same", or if for example some check "AtEnd" and some do not.

That's where Theories come into action.
They reduce many of the test-methods to one-liners, which prevents you from merging different aspects of testing.
Instead it enforces to have a precise test-statements which should be tested with different inputs.

Additionally it reports each single failing input instead of just failing at the first of them.

Is that something you would support?
It is not finished, and of course some scenarios have to be handled differently, we should not parse an entire file for each single test for example.

It's not really possible to check that request via git-compare at the moment, too many changes which are not recognized in the correct order.
If there is an interest in having this, I'd split this commit into single method-change-commits to enable a proper review.
And it is of course still a draft, it is not even fully finished for ParserTests.

Philipp-Binder committed:
refactor tests: ParseIdentifier()

rogusdev commented:
You don't need to switch to Theory to do what you are doing. It is very doable with XUnit to pass arguments to functions for various tests.

The problem with this change is about setup and teardown (ctor + dispose). If you separate the tests into multiple 1 liners, you run setup and teardown for each and every one of those, because that's the whole point, and that can add significant slowdowns to test running. In this set of tests, perhaps it won't be too noticeable, but there is a real chance that these changes could 3x-10x test runtime.

So, with these changes I would be concerned about showing things down annoyingly, and also unnecessarily dropping XUnit.

If you can address/justify those concerns, I would be open to this continuing.

Philipp-Binder commented:
> You don't need to switch to Theory to do what you are doing. It is very doable with XUnit to pass arguments to functions for various tests.

But with calling methods you won't get individual errors, since the test still fails at the first error.

> The problem with this change is about setup and teardown (ctor + dispose). If you separate the tests into multiple 1 liners, you run setup and teardown for each and every one of those, because that's the whole point, and that can add significant slowdowns to test running. In this set of tests, perhaps it won't be too noticeable, but there is a real chance that these changes could 3x-10x test runtime.

Sure, as I already stated we should not parse entire files for each individual test for example.
Where such expensive setup-actions are required, I need to find a solution addressing this concern, or the tests just stay as they were.
The huge advantage of individual errors is mainly about the direct (small) Parser tests, where you can directly see which cases are problematic and which are still green.

> So, with these changes I would be concerned about showing things down annoyingly, and also unnecessarily dropping XUnit.
> 
> If you can address/justify those concerns, I would be open to this continuing.

Does this address the concerns so far?
If your concern is only about performance, I guess there is no reason to stop until we see performance issues.
And if so, we can just leave these tests untouched.

Philipp-Binder commented:
Small addition:

I found some explanation about SharedContext in X-Unit, which could solve performance issues: https://xunit.net/docs/shared-context

I need to read through, but my first impression ist, that this can address all performance issues.

rogusdev commented:
Actually it turns out what you are doing is the XUnit way of doing this, I think I was thinking of nunit's approach, or some other framework.

So I support the path you are on. For sure this makes things easier to isolate for failures. And we can check performance at the end.

Thanks!

Philipp-Binder committed:
refactor tests: ParseOctalByte()

Philipp-Binder committed:
refactor tests: ParseOctalChar()

Philipp-Binder committed:
refactor tests: ParseHexByte()

Philipp-Binder committed:
refactor tests: ParseUtf8Char()

Philipp-Binder committed:
refactor tests: ParseUtf16Char()

Philipp-Binder committed:
refactor tests: ParseUtf32Char()

Philipp-Binder committed:
refactor tests: ParseEscapedChar()

Philipp-Binder committed:
refactor tests: ParseInterpolatedEnvVar()

Philipp-Binder committed:
refactor tests: ParseInterpolated()

Philipp-Binder committed:
refactor tests: ParseNotControlNorWhitespace()

Philipp-Binder committed:
refactor tests: ParseSpecialChar()

Philipp-Binder committed:
refactor tests: ParseComment()

Philipp-Binder committed:
refactor tests: ParseEmpty()

Philipp-Binder committed:
refactor tests: ParseUnquotedValue()

Philipp-Binder committed:
refactor tests: ParseDoubleQuotedValueContents()

Philipp-Binder committed:
refactor tests: ParseSingleQuotedValueContents()

Philipp-Binder committed:
refactor tests: ParseSingleQuotedValue()

Philipp-Binder committed:
remove unnecessary usings and unused property

Philipp-Binder commented:
Splitted to single commits for each changed method now, that way a review should be possible.

Updating the other methods will follow soon.

Philipp-Binder committed:
refactor tests: ParseDoubleQuotedValue()

Philipp-Binder committed:
refactor tests: TestExportExpression()

Philipp-Binder committed:
refactor tests: ParseValue()

Philipp-Binder committed:
refactor tests: ParseAssignment()

Philipp-Binder committed:
refactor tests: ParseDotenvFile

Philipp-Binder commented:
Ready with ParserTests, but searching for a better solution for ParseDotenvFile()-refactoring.
I think it has the advantage of separate failing, but it is not a really nice solution so far...

But at least the refactoring of this class can be reviewed, then I can use that feedback for refactoring the next classes.

Philipp-Binder committed:
refactor ParseDotenvFileShouldParseContents using MemberData combined with TheoryData

Philipp-Binder commented:
Found the solution for ParseDotenvFileShouldParseContents:

Using MemberData combined with TheoryData removes all these dictionaryRefs.
The solution is typesafe and you cannot miss any tests, compared with the loosely coupled DictionaryIndexing with InlineData.

It is still quite messy due to the long multiline strings, but that's also true for the old solution.
Maybe someone else has a suggestion to make that better readable?

Philipp-Binder committed:
refactor tests: LoadOptionsTests

Philipp-Binder commented:
While refactoring `LoadOptionsTests` I realized one real trade-off:

There is no good solution for showing which InlineData or TheoryData is under tests (no link back to the code-line).

While that is acceptable for most of the tests, there are cases (mainly with TheoryData I think) where this is quite inconvenient.
With LoadOptions it was quite horrible, because it is nearly impossible to see from the shown data (checking the compiled code is not a good way I guess :D ).

That's why I added a workaround, adding a new Type `IndexedTheoryData`.
There are some tradeoffs, the compiler cannot create exceptions for wrong parameters on the test method for example, but still the whole thing is typesafe etc. so that should not be a real problem (you get well explaining exceptions if you have wrong parameters).

But still it reduces the code so much, that this tradeoff should be okay imho.

Philipp-Binder committed:
refactor usage of tuples (use tuples only if they explain the data better)

Philipp-Binder committed:
refactor tests: ExtensionsTests

Philipp-Binder commented:
I just had a look through the other TestClasses now:

HelperTests and EnvReaderTests seem hard to rewrite, and 'm not sure how valuable this is.
1. I think this functionality is quite stable, there won't happen much
2. the focus should be on IConfiguration, which is the preferred way of ConfigurationData-Usage imho (IConfiguration allows to deserialize data into classes directly, no "manual" conversions)
3. all these methods do not work if you do not use SetEnvVars, which is a huge restriction (and an additional reason to use IConfiguration instead)

I would postpone EnvTests and EnvConfigurationTests, because there are potentially huge simplifications when working on #105 (decoupling Parsers and EnvVars) first.

What do you think so far?
I'm not completely happy with some of the code-constructs now, but in general I think it's  a good direction.

rogusdev commented:
Hey sorry for the delay, I did not realize you were waiting on me for next steps. Honestly, this is hard for me to confirm that you retained all existing tests correctly, but if this all passes, it looks broadly correct, and I do like the improvements. I would in fact encourage not doing more than this at this time and instead stop here and call it done, if indeed you are done with these changes.

So, I am fine with merging as-is if you are complete on this set of changes.

Philipp-Binder commented:
Not much time at the moment, and I'm still thinking about better solutions regarding some of the drawbacks...

Checking the PR is only possible commit by commit.
In general I just took the tests as-is and have them only re-written, but there are some special cases with changes or even completely gone.
I will comment those things as soon as I find some time again.

Philipp-Binder committed:
let IndexedTheoryData inherit from generic TheoryData;
declare TheoryData-lists as array of TheoryData<...> instead of IndexedTheoryData to make intellisense aware of the parameters;

Philipp-Binder commented:
At least I'm back, quite some other stuff at the moment ;)
I do some small cleanup now, making things better for intellisense and so on.

And here is the list/explanation for the dropped tests.
Please have a look and check if that's okay for you.

[53d04b8badd28845e7ff94e73faca3043575220c]
ParseUtf16Char --> Utf16CharShouldParseUntilEnd
`Assert.Equal("®", Encoding.Unicode.GetString(new byte[] { 0xae, 0x00 }));`
Seems to be more a test whether Unicode-Encoding works; test values need to be const now, so I dropped this one because the other tests do the same, just with a different notation.

[ba3c5834fba0b0669766be3b8752a65bd07d4156]
ParseUtf32Char --> Utf32CharShouldParseUntilEnd
```
Assert.Equal(RocketChar, Encoding.UTF32.GetString(new byte[] { 0x80, 0xf6, 0x01, 0x00 }));
Assert.Equal(RocketChar, Encoding.UTF32.GetString(new byte[] { 0x80, 0xf6, 0x1, 0x0 }));
```
Same thing as above.

[0629cf644297198d60e382f3738b293b167d9276]
ParseSpecialChar
```
// this caught a bug, once upon a time, where backslashes were
// getting processed by NCNW rather than EscapedChar inside SpecialChar
var parser = Parsers.SpecialChar
    .Or(Parsers.NotControlNorWhitespace("\"$"))
    .Or(Character.WhiteSpace.AtLeastOnce().Text());
Assert.Equal("\"", parser.AtEnd().Parse("\\\""));
```
This tests some "more elaborated" case, which should either be a separate test or should be dropped.

[12e7a2388042d2c5ef39518bb006f01107d9839d]
ParseAssignment
```
// Note that dotnet returns null if the env var is empty -- even if it was set to empty!
Assert.Null(Environment.GetEnvironmentVariable("EV_DNE"));
```
This just shows how the framework works. Either it should be an own test to assure that the framework is stable, or we should find a test which prooves that as a testcase for a project-method.

Philipp-Binder committed:
remove tuples in Theory data and add some hints as xml-doc instead; makes test-definition more precise/better readable;

Philipp-Binder commented:
Now I'd say this MR is ready, if you are okay with the above mentioned test-case drops.

rogusdev commented:
Can you make the "should be separate tests" actually separate? If so, that would make all of this totally good by me.

Philipp-Binder committed:
include previously dropped tests again

Philipp-Binder commented:
Included all of them again.

SpecialChar-Case got it's own fact, including the explanation that it caught a bug.

The other ones moved to `FrameworkTests`, since they are not checking actual code of this repository, but the surrounding framework.

Philipp-Binder commented:
I took a look at the test execution times now.
The number of tests is significantly higher of course, since every TheoryData counts as a single test. (from 68 to 282)
But the execution times are equal as far as I checked it, it is always around 3 minutes.

https://ci.appveyor.com/project/tonerdo/dotnet-env/history

rogusdev commented:
Sure, works for me. Thanks
